File: 3952-maximum-total-value-of-covered-indices/3952-maximum-total-value-of-covered-indices.kt
class Solution {
    fun maxTotal(nums: IntArray, s: String): Long {
        val neg = -1000000000000000000L

        var dp0 = 0L
        var dp1 = neg

        for (i in nums.indices) {
            var nxt0 = neg
            var nxt1 = neg

            if (s[i] == '0') {
                nxt0 = maxOf(dp0, dp1)
            } else {
                nxt1 = maxOf(dp0, dp1) + nums[i].toLong()

                if (i > 0) {
                    nxt0 = dp0 + nums[i - 1].toLong()
                }
            }

            dp0 = nxt0
            dp1 = nxt1
        }

        return maxOf(dp0, dp1)
    }
}

File: 3952-maximum-total-value-of-covered-indices/3952-maximum-total-value-of-covered-indices.rs
impl Solution {
    pub fn max_total(nums: Vec<i32>, s: String) -> i64 {
        let neg: i64 = -1_000_000_000_000_000_000;

        let mut dp0: i64 = 0;
        let mut dp1: i64 = neg;

        let chars = s.as_bytes();

        for i in 0..nums.len() {
            let mut nxt0: i64 = neg;
            let mut nxt1: i64 = neg;

            if chars[i] == b'0' {
                nxt0 = dp0.max(dp1);
            } else {
                nxt1 = dp0.max(dp1) + nums[i] as i64;

                if i > 0 {
                    nxt0 = dp0 + nums[i - 1] as i64;
                }
            }

            dp0 = nxt0;
            dp1 = nxt1;
        }

        dp0.max(dp1)
    }
}